File: one.java
package first;

import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

public class one {
    public static List<String> findSubstrings(String s) {
        List<String> result = new ArrayList<>();
        int n = s.length();
        int i = 0;

        while (i < n) {
            if (s.charAt(i) == '1') {
                int start = i;
                i++;


                boolean hasSecondOne = (i < n && s.charAt(i) == '1');
                if (hasSecondOne) {
                    i++;
                }

                int zeroCount = 0;
                while (i < n && s.charAt(i) == '0') {
                    zeroCount++;
                    i++;
                }

                if (zeroCount > 0 && i < n && s.charAt(i) == '1') {
                    result.add(s.substring(start, i + 1));
                    i++;
                } else {
                    i = start + 1;
                }
            } else {
                i++;
            }
        }
        return result;
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Введите строку: ");
        String inputString = scanner.nextLine();
        List<String> substrings = findSubstrings(inputString);

        System.out.println("Результат: " + String.join(", ", substrings));

        scanner.close();
    }
}

File: one.rb
def findSubstrings(s)
    result = []
    i = 0
    n = s.length
  
    while i < n
      if s[i] == '1'
        start = i
        pos = i + 1
        
        # Проверяем необязательную вторую '1'
        if pos < n && s[pos] == '1'
          pos += 1
        end
        
        # Ищем хотя бы один '0'
        zero_start = pos
        while pos < n && s[pos] == '0'
          pos += 1
        end
        
        # Проверяем что были нули и после них идет '1'
        if pos > zero_start && pos < n && s[pos] == '1'
          result << s[start..pos]
          i = pos + 1  # Перескакиваем после найденной подстроки
        else
          i += 1
        end
      else
        i += 1
      end
    end
  
    result
  end
  
  # Ввод строки от пользователя
  print "Введите строку: "
  input_string = gets.chomp
  
  # Поиск подстрок
  substrings = findSubstrings(input_string)
  

  puts "Результат: #{substrings.join(', ')}"
